### PR TITLE
refactor: refactor image blob to uri

### PR DESCRIPTION
**Pull request by bwanglzu**
1. make parameters `height`, `width` optional , interpret image shape from `blob`.
2. add grey scale support and `color_axis`.
3. more tests.

### CODE REVIEW

File: jina/types/document/__init__.py [hunk 1]
```diff
@@ -1007,14 +1007,13 @@ def convert_image_buffer_to_blob(self, color_axis: int = -1):
         self.blob = to_image_blob(io.BytesIO(self.buffer), color_axis)
 
     def convert_image_blob_to_uri(
-        self, width: int, height: int, resize_method: str = 'BILINEAR'
+        self, resize_method: str = 'BILINEAR', color_axis: int = -1
```
Comment (JoanFM) [inline jina/types/document/__init__.py:1010]: Is there a way to keep backward compatibility?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/document/__init__.py (hunk 1 above)
Comment (bwanglzu) [reply jina/types/document/__init__.py]: will keep `width` and `height`, won't deprecated it (only make this optional)

File: jina/types/document/__init__.py [hunk 2]
```diff
@@ -1007,14 +1007,23 @@ def convert_image_buffer_to_blob(self, color_axis: int = -1):
         self.blob = to_image_blob(io.BytesIO(self.buffer), color_axis)
 
     def convert_image_blob_to_uri(
-        self, width: int, height: int, resize_method: str = 'BILINEAR'
```
Comment (JoanFM) [inline jina/types/document/__init__.py:1010]: please add deprecation warning when they are used

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: jina/types/document/__init__.py (hunk 2 above)
Comment (bwanglzu) [reply jina/types/document/__init__.py:1010]: added:
DCFA7ED4-6E04-40EB-BE04-31F684287307

**Review by Anthony00107 — APPROVED** [review]

File: jina/types/document/__init__.py [hunk 3]
```diff
@@ -1006,15 +1006,25 @@ def convert_image_buffer_to_blob(self, color_axis: int = -1):
...
+        :param height: the height of the blob, if None, interpret from :attr:`blob` shape.
         :param resize_method: the resize method name
+        :param color_axis: the axis id of the color channel, ``-1`` indicates the color channel info at the last axis
+
+        ..note::
+            if both :attr:`width` and :attr:`height` were provided, will not resize. Otherwise, will get image size
+            by :attr:`self.blob` shape and apply resize method :attr:`resize_method`.
         """
-        png_bytes = png_to_buffer(self.blob, width, height, resize_method)
+        png_bytes = png_to_buffer(self.blob, width, height, resize_method, color_axis)
```
Comment (hanxiao) [inline jina/types/document/__init__.py:1027]: > if None, interpret from :attr:`blob` shape.

this does not implement this logic

File: jina/types/document/__init__.py [hunk 4]
```diff
@@ -1006,15 +1006,25 @@ def convert_image_buffer_to_blob(self, color_axis: int = -1):
         """
         self.blob = to_image_blob(io.BytesIO(self.buffer), color_axis)
 
+    @deprecated_parameter(params=['width', 'height'])
```
Comment (hanxiao) [inline jina/types/document/__init__.py:1009]: remove it

File: jina/helper.py [hunk 5]
```diff
@@ -141,6 +141,20 @@ def wrapper(*args, **kwargs):
     return deco
 
 
+def deprecated_parameter(params: List[str]):
```
Comment (hanxiao) [inline jina/helper.py:144]: remove this function

File: jina/types/document/__init__.py (hunk 4 above)
Comment (JoanFM) [reply jina/types/document/__init__.py:1009]: Why? This may be used by people in executors and solutions

**Review by JoanFM — APPROVED** [review]

**Pull request merged**